pymc3/distributions/multivariate.py

```diff
@@ -255,56 +262,53 @@ def random(self, point=None, size=None):
+        extra_dims = len(output_shape) - mu.ndim
+
+        # It was not a good idea to check mu.shape[:len(size)] == size,
+        # because it can get mixed among batch and event dimensions. Here, we explicitly chop off
+        # the size (sample_shape) and only broadcast batch and event dimensions.
+        if is_fast_drawable(self.mu) or point:
+            mu = mu.reshape((1,) * extra_dims + mu.shape)
+        else:
+            mu = mu.reshape(size + (1,) * extra_dims + mu.shape[len(size) :])
 
-        if self._cov_type == "cov":
-            mu, cov = draw_values([self.mu, self.cov], point=point, size=size)
-            if mu.shape[-1] != cov.shape[-1]:
-                raise ValueError("Shapes for mu and cov don't match")
+        mu = np.broadcast_to(mu, output_shape)
 
-            try:
-                dist = stats.multivariate_normal(mean=mu, cov=cov, allow_singular=True)
-            except ValueError:
-                size += (mu.shape[-1],)
-                return np.nan * np.zeros(size)
-            return dist.rvs(size)
-        elif self._cov_type == "chol":
-            mu, chol = draw_values([self.mu, self.chol_cov], point=point, size=size)
-            if size and mu.ndim == len(size) and mu.shape == size:
-                mu = mu[..., np.newaxis]
-            if mu.shape[-1] != chol.shape[-1] and mu.shape[-1] != 1:
-                raise ValueError("Shapes for mu and chol don't match")
-            broadcast_shape = np.broadcast(np.empty(mu.shape[:-1]), np.empty(chol.shape[:-2])).shape
-
-            mu = np.broadcast_to(mu, broadcast_shape + (chol.shape[-1],))
-            chol = np.broadcast_to(chol, broadcast_shape + chol.shape[-2:])
-            # If mu and chol were fixed by the point, only the standard normal
-            # should change
-            if mu.shape[: len(size)] != size:
-                std_norm_shape = size + mu.shape
-            else:
-                std_norm_shape = mu.shape
-            standard_normal = np.random.standard_normal(std_norm_shape)
+        if mu.shape[-1] != param.shape[-1]:
+            raise ValueError(f"Shapes for mu and {self._cov_type} don't match")
+
+        if self._cov_type == "cov":
+            # Calculate mu + L * samples, L = chol(cov)
+            chol = linalg.cholesky(param, lower=True)
+            standard_normal = np.random.standard_normal(output_shape)
             return mu + np.einsum("...ij,...j->...i", chol, standard_normal)
+        elif self._cov_type == "chol":
+            standard_normal = np.random.standard_normal(output_shape)
+            return mu + np.einsum("...ij,...j->...i", param, standard_normal)
         else:
-            mu, tau = draw_values([self.mu, self.tau], point=point, size=size)
-            if mu.shape[-1] != tau[0].shape[-1]:
-                raise ValueError("Shapes for mu and tau don't match")
-
-            size += (mu.shape[-1],)
+            # Calculate mu + L^{-T} * samples, L = chol(tau)
```

Comment:
I'll have to review the math here a bit before being able to comment on the conditional branch

Reply:
After reviewing [the math](http://www.statsathome.com/2018/10/19/sampling-from-multivariate-normal-precision-and-covariance-parameterizations/) a bit. I think that this conditional branch needs major changes.

Reply:
There was a very interesting derivation about this on stack exchange. Let me hunt for it again and give a mention here.

Reply:
This is the stack exchange [link](https://math.stackexchange.com/questions/1845132/sample-points-from-a-multivariate-normal-distribution-using-only-the-precision-m) I am referring to. Also, the code written is equivalent to the equation (for precision matrix) mentioned in the link shared above

![CodeCogsEqn](https://user-images.githubusercontent.com/43073325/99538808-7b653280-29d3-11eb-89f4-a2a19c477b7c.gif)

And this $L_{\Omega}^{-T}  Z$ is obtained by back-solving.

Reply:
Now, this `tau` conditional branch is corrected, first by inverting the precision matrix to get `cov` and then taking cholesky decomposition using `np.linalg` instead of `scipy.linalg`. I used `np.linalg.cholesky` and `np.linalg.inv` because they can operate on a stack of 2-d matrices, thereby making our job easier when `tau` also has sample shape dimensions.